juju/charm.py

```diff
@@ -82,19 +117,145 @@ def __init__(self, framework, key, metadata):
+    the ``requires``, ``provides``, and ``peers`` :class:`RelationDefinition`_
+    items.  If needed, the role of the relation definition can be obtained from
+    its ``role`` attribute.
+    """
+    def __init__(self, raw=None):
+        raw = raw or {}
+        self.name = raw.get('name', '')
+        self.summary = raw.get('summary', '')
+        self.description = raw.get('description', '')
+        maintainers = []
+        if 'maintainer' in raw:
+            maintainers.append(raw['maintainer'])
+        if 'maintainers' in raw:
+            maintainers.extend(raw['maintainers'])
+        self.maintainers = tuple(maintainers)
+        self.tags = tuple(raw.get('tags', []))
+        self.terms = tuple(raw.get('terms', []))
+        self.series = tuple(raw.get('series', []))
+        self.subordinate = raw.get('subordinate', False)
+        self.min_juju_version = raw.get('min-juju-version')
+        self.requires = {name: RelationDefinition('requires', name, rel)
+                         for name, rel in raw.get('requires', {}).items()}
+        self.provides = {name: RelationDefinition('provides', name, rel)
+                         for name, rel in raw.get('provides', {}).items()}
+        self.peers = {name: RelationDefinition('peers', name, rel)
+                      for name, rel in raw.get('peers', {}).items()}
+        self.relations = {}
+        self.relations.update(self.requires)
+        self.relations.update(self.provides)
+        self.relations.update(self.peers)
+        self.storage = {name: StorageDefinition(name, store)
+                        for name, store in raw.get('storage', {}).items()}
+        self.resources = {name: ResourceDefinition(name, res)
+                          for name, res in raw.get('resources', {}).items()}
+        self.payloads = {name: PayloadDefinition(name, payload)
+                         for name, payload in raw.get('payloads', {}).items()}
+        self.extra_bindings = tuple(raw.get('extra-bindings', {}))
+
+
+class RelationDefinition:
+    """Object containing metadata about a relation definition.
+
+    Supported fields are:
+
+        * ``role``
+        * ``relation_name``
+        * ``interface_name``
+        * ``scope``
+    """
+    def __init__(self, role, relation_name, raw):
+        self.role = role
+        self.relation_name = relation_name
+        self.interface_name = raw['interface']
+        self.scope = raw.get('scope')
+
+
+class StorageDefinition:
+    """Object containing metadata about a storage definition.
+
+    Supported fields are:
+    """
+    def __init__(self, name, raw):
+        self.storage_name = name
+        self.type = raw['type']
+        self.description = raw.get('description', '')
+        self.shared = raw.get('shared', False)
+        self.read_only = raw.get('read-only', False)
+        self.minimum_size = raw.get('minimum-size')
+        self.location = raw.get('location')
+        self.is_multiple = 'multiple' in raw
```

Comment:
What is that about?  I don't recall seeing the details for it before.

Either way, we need tests for parts like these, which are more than just a trivial get. Ideally we'll eventually get tests for all of those, but the more complex ones we need sooner rather than later.

Reply:
+1 to tests if we stick with this logic.  However, I wasn't entirely confident about translating the `multiple` block from the below to the `is_multple` and `range` tuple.  It feels more Pythonic, but it deviates from what the charm author wrote.  So maybe I should make it more of a simple mapping and then we might not need tests.

```yaml
storage:
  foo:
    type: block
    multiple:
      range: 0-10
```